lib/src/client/dio_interceptor.dart

```diff
@@ -0,0 +1,226 @@
+      return handler.next(response);
+    } catch (error, stackTrace) {
+      print('[x] resendRequest error');
+      throw RequestsSignatureException('ResendERR\n$error\n$stackTrace');
+    }
+  }
+
+  /// Signs the outgoing request with a request signature.
+  Future<void> _signRequest(RequestOptions options) async {
+    print('[i] _signRequest called');
+    try {
+      print('[✓] _signRequest triggered');
+      options.headers.remove(_options.headerName);
+
+      final signatureBodySourceComponents =
+          _options.signatureBodySourceComponents.isNotEmpty
+              ? _options.signatureBodySourceComponents
+              : DefaultConstants.signatureBodySourceComponents;
+
+      Uint8List? body;
+      // Encode request body if required by signature components
+      if (options.data != null &&
+          signatureBodySourceComponents
+              .contains(SignatureBodySourceComponents.body)) {
+        body = utf8.encode(options.data.toString());
+      }
+      print('[i] _signReq Header: ${options.headers[HttpHeaders.dateHeader]}');
+
+      // Build parameters for constructing the signature body
+      final signatureBodySourceParameters = SignatureBodySourceParameters(
+          options.method,
+          options.uri,
+          options.headers
+              .map((key, value) => MapEntry(key.toString(), value.toString())),
+          _uuid.v4(), // Generate a nonce
+          _getTimestamp(options.headers[HttpHeaders.dateHeader]),
+          _options.clientId!,
+          signatureBodySourceComponents,
+          body: body);
+
+      // Build the signature body source based on parameters
+      final signatureBodySource = await _signatureBodySourceBuilder
+          .build(signatureBodySourceParameters);
+
+      // Construct parameters for creating the signature
+      final signatureBodyParameters =
+          SignatureBodyParameters(signatureBodySource, _options.clientSecret!);
+
+      // Generate the signature using the signer
+      final signatureBody =
+          await _signatureBodySigner.sign(signatureBodyParameters);
+
+      // Format the signature header based on the specified pattern
+      final signatureHeader = _options.signaturePattern
+          .replaceAll("{ClientId}", _options.clientId!)
+          .replaceAll("{Nonce}", signatureBodySourceParameters.nonce)
+          .replaceAll(
+              "{Timestamp}", signatureBodySourceParameters.timestamp.toString())
+          .replaceAll("{SignatureBody}", signatureBody);
+
+      print('[DEBUG] Generated Signature Header: $signatureHeader');
+
+      // Add the signature header to the request headers
+      options.headers[_options.headerName] = signatureHeader;
+    } catch (error, stackTrace) {
+      print('[x] _signRequest error');
+      throw RequestsSignatureException('$error\n$stackTrace');
+    }
+  }
+
+  // Returns what the client thinks is the current time.
+  static int getTime(DateTime? timestamp) {
```

Comment:
This method is still different than the .NET implementation. You need to implement the inversion of control to have the option to control the current time in unit tests.

Reply:
This is where things get a little but messy, .NET and Dart are different and can't be replicated as it is, so this is just a work around

Reply:
I understand that you may not have access to a `Dictionary<string, object>` in which you can pass a function for each request like in .NET. However, you could pass an optional function in the constructor of the dio interceptor.

Reply:
+update:
alternative found